Userland/mount.cpp

```diff
@@ -96,11 +96,23 @@ bool mount_all()
             continue;
         }
 
-        dbg() << "Mounting " << devname << "(" << fstype << ")"
+        const char* filename = parts[0].characters();
+        int fd = open(filename, O_RDWR);
+        if (fd < 0)
+            fd = open(filename, O_DIRECTORY);
```

Comment:
@bugaevc what do you think about this pattern? is it a good thing to try to open a path with other flag if the first attempt failed?

Reply:
The pattern is fine, but the specific code is sadly wrong.
* You should always pass either `O_RDONLY`, `O_WRONLY`, `O_EXEC`, `O_RDWR` (or a combination of them). You can add other flags, but one of those two should always be present.
* `O_DIRECTORY`, unlike the mode flags I listed above, is not a special mode to open directories, it's a special flag that modifies `open()` behavior. Specifically, it asks `open()` to fail if the thing you're opening turns out not to be a directory. This is *not* what you want here.

Reply:
`fd = open(filename, O_RDONLY | O_DIRECTORY);` - can I do that in the second attempt?

Reply:
That would fix my point number 1, but not my point number 2 🙂.
Just `open(filename, O_RDONLY)`

Reply:
but then we could have a problem with read-only file descriptors, so if you try to write to it, it will fail in the kernel, right?

Reply:
Yeah, you pass a r/o fd to `mount()` and that may fail (in `sys$mount()` or later), or it may be fine (r/o should be enough for bind-mounts). But it's not like passing `O_DIRECTORY` was going to make your fd read-write :slightly_smiling_face: — we'd be hitting this code path exactly when we cannot open the file as read-write, so we try read-only as a fallback.

Reply:
OK, thanks for the help :)